alloc-customization-point.html

```diff
@@ -165,11 +201,100 @@ <h2>About the customization points, request for feedback/decision</h2>
+
+ <h3>Class template domain_allocator_traits</h3>
+<code>namespace std {
+         template &lt;class T, class Alloc&gt; struct domain_allocator_traits {};
+         }
+         </code>
+</p></blockquote>
+<p><tt>domain_allocator_traits</tt> provides a means to get uses-allocator construction with allocator types other than the
+type specific allocator. Meets the <i>Cpp17BinaryTypeTrait</i> requirements ([meta.rqmts]). </p>
+
+<p>If <tt>domain_allocator_traits&lt;class T, class Alloc&gt;::value</tt> is <tt>true</tt>,  there will exist a 
+function object member of <tt>domain_allocator_traits</tt> called <tt>rebind</tt> which when invoked with a parameter of type Alloc
+will return type DomainAlloc, and type T will have trailing_allocator convention constructors using DomainAlloc. 
+Otherwise, <tt>domain_allocator_traits&lt;class T, class Alloc&gt;::value</tt> will be <tt>false</tt>. </p>
+
+<p>Implementation shall provide a definition of <tt>domain_allocator_traits&lt;class T, class Alloc&gt</tt> that is 
+derived from <tt>true_­type</tt> if there exists a function object member <tt>T::domain_alloc_rebind</tt> for 
+which <tt>is_invocable&lt;decltype(T::domain_alloc_rebind), const _Alloc&&gt;::value</tt> is <tt>true</tt>. </p> 
+
+
+
+<p><br/>Modify 20.10.8.1 uses_­allocator trait [allocator.uses.trait]</p>
+
+<p><tt>template &lt;class T, class Alloc&gt; struct uses_allocator;</tt>
+<br/>
+<i>#Remarks:</i> Automatically detects whether <tt>T</tt> has a nested allocator_­type that is convertible from <tt>Alloc</tt>. 
+Meets the <i>Cpp17BinaryTypeTrait</i> requirements ([meta.rqmts]). The implementation shall provide a definition that is 
+derived from <tt>true_­type</tt> if<ins>
+<br/> - </ins>the <i>qualified-id</i> <tt>T::allocator_­type</tt> is valid and denotes a type ([temp.deduct]) and 
+<tt>is_­convertible_­v&lt;Alloc, T::allocator_­type&gt; != false</tt>,<ins> or
+<br/> - if <tt>__domain_allocator_traits&lt;T, Alloc&gt;::value</tt> is <tt>true</tt></ins> 
+<br/><br/>
+<del>otherwise it</del><ins>It</ins> shall be derived from <tt>false_­type</tt><ins> otherwise</ins>. 
+A program may specialize this template to derive from <tt>true_­type</tt> for a program-defined type <tt>T</tt> that does not
+have a nested <tt>allocator_­type</tt> but nonetheless can be constructed with an allocator where either:
+<br/>- the first argument of a constructor has type allocator_­arg_­t and the second argument has type <tt>Alloc</tt> or
+<br/>- the last argument of a constructor has type <tt>Alloc</tt>.
+
+</p>
+
+
+<p>
+Modify [allocator.uses.construction]/p1
+<blockquote>
+ <p>
+<i>Uses-allocator construction</i> with allocator <tt>Alloc</tt> and constructor arguments <tt>args...</tt> refers to the construction
+of an object of type <tt>T</tt> such that <tt>Alloc</tt> is passed to the constructor of T if T uses an allocator type compatible
+with <tt>Alloc</tt><ins>  or <tt>__domain_allocator_traits&lt;T, Alloc&gt;::value</tt> is <tt>true</tt></ins>.
```

Comment:
This paragraph explains what uses-allocator construction means. We should add "or an allocator compatible with the result of invoking domain_allocator_traits<T, Alloc>::rebind with the allocator." instead of adding dependency on the value of the trait.

Reply:
I see what you mean. How about
"Uses-allocator construction with allocator Alloc and constructor arguments args... refers to the construction of an object of type T such that Alloc or the result of  invoking domain_allocator_traits<T, Alloc>::rebind with Alloc is passed to the constructor of T if T uses an allocator type compatible with Alloc."

Should we also say what compatible means ?
I'm also bothered by the type Alloc is both type and an object in the above statement. Suggestions ?

Reply:
The suggestion is fine. The concerns are pre-existing, so I'm not too concerned about them. "An argument of type Alloc" might work for the type/object concern.

Reply:
Modified to "Uses-allocator construction with allocator alloc of type Alloc and constructor arguments args... refers to the construction of an object of type T such that alloc or the result of invoking domain_allocator_traits<T, Alloc>::rebind with alloc is passed to the constructor of T if T uses an allocator type compatible with Alloc."